aiohttp/helpers.py

```diff
@@ -103,6 +104,177 @@ def encode(self):
         return 'Basic %s' % base64.b64encode(creds).decode(self.encoding)
 
 
+def parse_pair(pair):
+    key, value = pair.split('=', 1)
+
+    # If it has a trailing comma, remove it.
+    if value[-1] == ',':
+        value = value[:-1]
+
+    # If it is quoted, then remove them.
+    if value[0] == value[-1] == '"':
+        value = value[1:-1]
+
+    return key, value
+
+
+def parse_key_value_list(header):
+    return {
+        key: value for key, value in
+        [parse_pair(header_pair) for header_pair in header.split(' ')]
+    }
+
+
+class DigestAuth():
+    """HTTP digest authentication helper.
+    The work here is based off of
+    https://github.com/requests/requests/blob/v2.18.4/requests/auth.py.
+    """
+
+    def __init__(self, username, password, session, previous=None):
+        if previous is None:
+            previous = {}
+
+        self.username = username
+        self.password = password
+        self.last_nonce = previous.get('last_nonce', '')
+        self.nonce_count = previous.get('nonce_count', 0)
+        self.challenge = previous.get('challenge')
+        self.args = {}
+        self.session = session
+
+    async def request(self, method, url, *, headers=None, **kwargs):
+        if headers is None:
+            headers = {}
+
+        # Save the args so we can re-run the request
+        self.args = {
+            'method': method,
+            'url': url,
+            'headers': headers,
+            'kwargs': kwargs
+        }
+
+        if self.challenge:
+            headers[hdrs.AUTHORIZATION] = self._build_digest_header(
+                method.upper(), url
+            )
+
+        response = await self.session.request(
+            method, url, headers=headers, **kwargs
+        )
+
+        # Only try performing digest authentication if the response status is
+        # from 400 to 500.
+        if 400 <= response.status < 500:
```

Comment:
In one of my tests, when username-password pair was incorrect, this led to infinite recursion. I think the fix is to call `_handle_401` only if `self.challenge` is `None`.